### ISSUE TITLE
Wrong variation menu order with disabled variations

### ISSUE THREAD

**Issue by jorisvst** [woothemes/woocommerce#32833]
### Prerequisites

- [x] I have carried out troubleshooting steps and I believe I have found a bug.
- [X] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

When there is a variable product with multiple variations and some variations are not enabled, the menu_order is not correctly saved when only one variation is saved.

This happens because the function `WC_Product_Data_Store_CPT::sort_all_product_variations`, only updates the menu_order of variations with `post_status = 'publish'`.

This function `WC_Product_Data_Store_CPT::sort_all_product_variations` is being called in the function `WC_Meta_Box_Product_Data::save_variations` where it handles the variations data when clicking on 'save changes'.

So when you change a value (e.g. the price value) of only one variation and click on 'save changes', the menu_order of all variations with `post_status = 'publish'` are being changed with the function `WC_Product_Data_Store_CPT::sort_all_product_variations`. The for loop then updates the prop menu_order of the changed variation.
This does not happens when you drag and drop a variation because then, all variations are being updated in the for loop resulting to correctly update the menu_order prop after the `WC_Meta_Box_Product_Data::save_variations` changed them wrongly.

### Expected behavior

When changing the value of one variations (e.g. the price value), the menu_order of the variations should not change when there are disabled variations.

### Actual behavior

The menu_order changes when there are disabled variations because the function `WC_Product_Data_Store_CPT::sort_all_product_variations` only saves the menu_order of variations with post_status = 'publish'.

### Steps to reproduce

1. Create a new variable product with 4 variations.
2. Disable the first and third variation by unchecking the checkbox 'enabled'.
3. Save changes.
4. Add a price for the second variation.
5. Save changes.
6. When saved and variations are reloaded with AJAX, the third and fourth variation switches from position while no sorting was changed.

<details>
<summary>WordPress Environment</summary>
### WordPress Environment ###

WordPress address (URL): -
Site address (URL): -
WC Version: 6.4.1
REST API Version: ✔ 6.4.1
WC Blocks Version: ✔ 7.2.2
Action Scheduler Version: ✔ 3.4.0
WC Admin Version: ✔ 3.3.2
Log Directory Writable: ✔
WP Version: 5.9.3
WP Multisite: –
WP Memory Limit: 1 GB
WP Debug Mode: ✔
WP Cron: ✔
Language: nl_BE
External object cache: –

### Server Environment ###

Server Info: Apache/2
PHP Version: 7.4.28
PHP Post Max Size: 128 MB
PHP Time Limit: 30
PHP Max Input Vars: 1000
cURL Version: 7.71.0
OpenSSL/1.1.1d

SUHOSIN Installed: –
MySQL Version: 10.4.18-MariaDB-cll-lve
Max Upload Size: 128 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 6.4.1
WC Database Prefix: at_
Total Database Size: 214.66MB
Database Data Size: 121.76MB
Database Index Size: 92.90MB
at_woocommerce_sessions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
at_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
at_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
at_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.17MB + Engine InnoDB
at_woocommerce_order_items: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
at_woocommerce_order_itemmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
at_woocommerce_tax_rates: Data: 0.02MB + Index: 0.08MB + Engine InnoDB
at_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
at_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
at_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
at_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
at_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
at_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
at_woocommerce_log: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
at_actionscheduler_actions: Data: 1.02MB + Index: 0.25MB + Engine InnoDB
at_actionscheduler_claims: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
at_actionscheduler_groups: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
at_actionscheduler_logs: Data: 1.03MB + Index: 1.11MB + Engine InnoDB
at_cartbounty: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
at_commentmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
at_comments: Data: 0.30MB + Index: 0.09MB + Engine InnoDB
at_links: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
at_mollie_pending_payment: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
at_options: Data: 6.03MB + Index: 1.08MB + Engine InnoDB
at_pmxe_exports: Data: 0.03MB + Index: 0.00MB + Engine InnoDB
at_pmxe_google_cats: Data: 0.39MB + Index: 0.00MB + Engine InnoDB
at_pmxe_posts: Data: 0.23MB + Index: 0.00MB + Engine InnoDB
at_pmxe_templates: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
at_pmxi_files: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
at_pmxi_hash: Data: 3.13MB + Index: 0.00MB + Engine InnoDB
at_pmxi_history: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
at_pmxi_images: Data: 0.08MB + Index: 0.00MB + Engine InnoDB
at_pmxi_imports: Data: 1.50MB + Index: 0.00MB + Engine InnoDB
at_pmxi_posts: Data: 2.52MB + Index: 0.00MB + Engine InnoDB
at_pmxi_templates: Data: 0.34MB + Index: 0.00MB + Engine InnoDB
at_postmeta: Data: 33.23MB + Index: 15.30MB + Engine InnoDB
at_posts: Data: 12.34MB + Index: 3.58MB + Engine InnoDB
at_redirection_404: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
at_redirection_groups: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
at_redirection_items: Data: 2.52MB + Index: 3.47MB + Engine InnoDB
at_redirection_logs: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
at_searchwp_index: Data: 14.52MB + Index: 28.09MB + Engine InnoDB
at_searchwp_log: Data: 0.14MB + Index: 0.22MB + Engine InnoDB
at_searchwp_status: Data: 0.34MB + Index: 0.48MB + Engine InnoDB
at_searchwp_tokens: Data: 1.52MB + Index: 2.52MB + Engine InnoDB
at_termmeta: Data: 0.11MB + Index: 0.14MB + Engine InnoDB
at_terms: Data: 0.08MB + Index: 0.09MB + Engine InnoDB
at_term_relationships: Data: 4.05MB + Index: 1.28MB + Engine InnoDB
at_term_taxonomy: Data: 0.11MB + Index: 0.09MB + Engine InnoDB
at_usermeta: Data: 30.52MB + Index: 29.09MB + Engine InnoDB
at_users: Data: 2.52MB + Index: 3.52MB + Engine InnoDB
at_wc_admin_notes: Data: 0.06MB + Index: 0.00MB + Engine InnoDB
at_wc_admin_note_actions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
at_wc_category_lookup: Data: 0.05MB + Index: 0.00MB + Engine InnoDB
at_wc_customer_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
at_wc_download_log: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
at_wc_order_coupon_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
at_wc_order_product_lookup: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
at_wc_order_stats: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
at_wc_order_tax_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
at_wc_product_attributes_lookup: Data: 0.11MB + Index: 0.06MB + Engine InnoDB
at_wc_product_meta_lookup: Data: 0.50MB + Index: 0.73MB + Engine InnoDB
at_wc_rate_limits: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
at_wc_reserved_stock: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
at_wc_tax_rate_classes: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
at_wc_webhooks: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
at_wdr_order_discounts: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
at_wdr_order_item_discounts: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
at_wdr_rules: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
at_yoast_indexable: Data: 1.52MB + Index: 0.47MB + Engine InnoDB
at_yoast_indexable_hierarchy: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
at_yoast_migrations: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
at_yoast_primary_term: Data: 0.06MB + Index: 0.03MB + Engine InnoDB
at_yoast_seo_links: Data: 0.02MB + Index: 0.03MB + Engine InnoDB

### Post Type Counts ###

acf-field: 3
acf-field-group: 2
attachment: 2318
cf7_entry: 1
gardenbook: 1378
nav_menu_item: 129
page: 34
product: 1205
product_variation: 2699
revision: 3295
shop_order: 7
templatera: 2
wcpf_item: 23
wcpf_project: 2
wishlist: 23
wpcf7_contact_form: 2

### Security ###

Secure connection (HTTPS): ❌
					Your store is not using HTTPS. Learn more about HTTPS and SSL Certificates.
Hide errors from visitors: ❌Error messages should not be shown to visitors.

### Active Plugins (1) ###

WooCommerce: by Automattic – 6.4.1

### Inactive Plugins (43) ###

Admin category search: by Atilla Ordog – 1.2
Advanced Custom Fields PRO: by Delicious Brains – 5.12.2
BackWPup: by Inpsyde GmbH – 3.10.0
Bulk remove posts from category: by MasterNs – 3.2.1
CartBounty - Save and recover abandoned carts for WooCommerce: by Streamline.lv – 7.1.2.5
Classic Widgets: by WordPress Contributors – 0.3
CMB2: by CMB2 team – 2.10.1
Contact Form 7: by Takayuki Miyoshi – 5.5.6
Contact Form 7 Storage: by Kaspars Dambis – 2.0.3
Duplicate Page: by mndpsingh287 – 4.4.8
Europabank MPI: by Europabank NV. – 1.0
File Media Renamer: by Bradley B. Dalina – 1.2
iSide Cookie Banner: by iSide – 1.7.2
Jetpack: by Automattic – 10.8
Mollie Payments for WooCommerce: by Mollie – 7.0.4
MultiSafepay: by MultiSafepay – 4.13.1
NinjaFirewall (WP Edition): by The Ninja Technologies Network – 4.5.1
Product Filters for WooCommerce: by WooCommerce – 1.2.5
Redirection: by John Godley – 5.2.3
Regenerate Thumbnails: by Alex Mills (Viper007Bond) – 3.1.5
SearchWP: by SearchWP – 4.2.2
SearchWP Live Ajax Search: by SearchWP
LLC – 1.6.2

SearchWP Redirects: by SearchWP – 1.4.2
Simple Comment Editing: by Ronald Huereca – 2.7.1
Simple Custom Post Order: by Colorlib – 2.5.6
Templatera: by WPBakery – 2.0.4
Trustbadge Reviews for WooCommerce: by vendidero – 4.0.13
User Role Editor: by Vladimir Garagulya – 4.61.2
W3 Total Cache: by BoldGrid – 2.2.1
WooCommerce Conditional Shipping and Payments: by WooCommerce – 1.12.0
WooCommerce EU VAT Number: by WooCommerce – 2.4.1
WooCommerce Google Analytics Integration: by WooCommerce – 1.5.8
WooCommerce WishLists: by Element Stark – 2.2.6
Woo Discount Rules: by Flycart – 2.4.0
Woo Discount Rules PRO 2.0: by Flycart – 2.4.0
WP All Export - WooCommerce Export Add-On Pro: by Soflyy – 1.0.4
WP All Export Pro: by Soflyy – 1.7.5
WP All Import - User Import Add-On Pro: by Soflyy – 1.1.6
WP All Import - WooCommerce Import Add-On Pro: by Soflyy – 3.3.0
WP All Import Pro: by Soflyy – 4.7.2
WPBakery Page Builder: by Michael M - WPBakery.com – 6.9.0
WP Mail SMTP: by WPForms – 3.3.0
Yoast SEO: by Team Yoast – 18.6

### Dropin Plugins (1) ###

advanced-cache.php: advanced-cache.php

### Must Use Plugins (1) ###

iside-mustuse.php: by iSide – 1.0

### Settings ###

API Enabled: –
Force SSL: –
Currency: EUR (€)
Currency Position: left
Thousand Separator: .
Decimal Separator: ,
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: –

### WC Pages ###

Shop base: #9 - /tuincentrum/
Cart: #10 - /winkelmand/
Checkout: #11 - /afrekenen/
My account: ❌ Page does not contain the [woocommerce_my_account] shortcode.
Terms and conditions: ❌ Page not set

### Theme ###

Name: Twenty Twenty-Two
Version: 1.0 (update to version 1.1 is available)
Author URL: https://wordpress.org/
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ✔

### Templates ###

Overrides: –

### Action Scheduler ###

Complete: 583
Oldest: 2022-02-05 15:26:40 +0100
Newest: 2022-04-28 14:20:58 +0200

Pending: 79
Oldest: 2022-04-28 14:20:59 +0200
Newest: 2022-04-29 09:16:14 +0200

Failed: 2
Oldest: –
Newest: –

### Status report information ###

Generated at: 2022-04-29 11:14:22 +02:00
`
</details>

### Isolating the problem

- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [x] I can reproduce this bug consistently using the steps above.

**Comment by woocommercebot** [comment]
We are adding the `status: needs reproduction` label to this issue to try reproduce it on the   current released version of WooCommerce.

  Thank you for your patience.

**Comment by swatipawarGS** [comment]
Hi @joris-iSide,

Thank you for taking the time to report this bug, we really appreciate your help. We are able to reproduce the issue on our end using WooCommerce 6.4.1 version and WordPress 5.9.3 version. 

Please find below screencast for reference :

https://user-images.githubusercontent.com/94531721/166220550-5fe87922-bb86-421c-85d8-13b65af39fbe.mp4

Priority of this issue will be assigned shortly after we’d go through a round of prioritization. Please note it may take a couple of days.

**Comment by woocommercebot** [comment]
We are adding the `status: needs reproduction` label to this issue to try reproduce it on the   current released version of WooCommerce.

  Thank you for your patience.

**Comment by peterfabian** [comment]
Thanks for taking the time and reporting the bug. Would you like to try submit a PR with a solution, since you tracked down the problem rather well?

**Comment by jorisvst** [comment]
Hi @peterfabian. 
A [PR](https://github.com/woocommerce/woocommerce/pull/36343) has been submitted that fixes this problem. 

We applied this solution to a customer of us but are not 100% sure that this is a complete fix.
In the case of this customer it is a full working solution, however we did not take all possible cases into account yet, so it might be possible that it creates a conflict in a special situation, although I think that will be a small chance.

I’d like to hear if some improvements can be made.
Thanks!
